### ISSUE TITLE
`undefined` is acting like `any` in generics extension checks

### ISSUE THREAD

**Issue by MarkKahn** [microsoft/TypeScript#49862]
# Bug Report

### 🔎 Search Terms

generic function call

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about generics

### ⏯ Playground Link

[playground link](https://www.typescriptlang.org/play?target=99&ts=4.8.0-beta&ssl=30&ssc=1&pln=30&pc=80#code/C4TwDgpgBAkgZgQQHYgDwBUB8UC8UAMUEAHsBEgCYDOUAFAIxQBkU6AlFAPxTABOArtABcUJBABuEXgChpoSD2QgoKlXnhLUAQxSYA3KsMB6I1ADGW-gHMAFsDnhowAKpIA1kgD2AdyS5YiCio-O5evvqqJqKewOaWtvby0HCenoZqAZohFBBwAJZiFBFRXrEW1nayOWYANlq8ySFmwHmefgBGqTUYRKTk1JlBWJi04lo1gpwi7CIA5OUJs3rS1XUNUHBNLW1QnZ7dWKPjk9Nsc6VxFcBLVRC19Y1Iza0dXQBMPSRklDQA3gDasyseXa7SkeSoNlmAF1pgBfEZjCYQKasM5QebxOw3VYPDZbF67d4YRHHFGnc4xS6LZbSMxtKixBD+PY1WgAFjeUC0NB0IDYy3pSEZUAAQiyuhyuTyoCEPD4kAK6QzYgBhCX7WizWZKoUigAiGrZ+F1KqgAFEjVLTcLYgAxK2zGwQGo1Tw6wVmgDiVr4ghtIoAEla4OMqBAA7EYFakPxXZGoAApK3ZXIFCAUAVQKILOwbTy8KBUTwAW2gDR5bU4nGVtqgAGkrf9oQmADJWt6Kz11gCyVvwXdketiADkjW9aKn8oUszmsbFqzXh1AAPLj620pJFnR5UCq51mNz+Kfpii9b4DAFAkFg3gQqGw2WUNOFOFcDEgCBUWZQSlLIA)

### 💻 Code

```ts
type IfAny<T> = 0 extends (1 & T) ? true : never

declare function bool<T extends IfAny<T>>(value?: T): 'caught';
declare function bool<T>(value?: T): 'not caught';

declare function bool2<T extends {['gibberish']: T}>(value?: T): 'caught';
declare function bool2<T>(value?: T): 'not caught';

const A = bool(undefined);  // caught ???
const B = bool2(undefined); // caught ???
const C = bool();           // caught ???
const D = bool2();          // caught ???
```

### 🙁 Actual behavior

both implicit and explicit `undefined` arguments seems to act like `any` when used in a generic when the argument is optional.  This behavior is not seen when the argument is required

### 🙂 Expected behavior

`undefined` shouldn't extend everything

**Comment by MartinJohns** [comment]
Your function is typed `T | undefined` because it's an optional parameter. So when you call it passing `undefined` or without an argument, there's nothing to infer `T` from, so `T` resolves to `never`, and `never` extends every type.

**Comment by MarkKahn** [comment]
> never extends every type.

This seems incredibly!  What's the logic here?

**Comment by fatcerberus** [comment]
`never` is the bottom type which is the opposite of the top type `unknown`.  It conceptually represents the result of operations that never complete (a function that throws, e.g.), so you can treat it as any type you want.  In set-theoretical terms, `never` is an empty set (because there are no values of that type)--which is a subset of every other set.
